energy-analysis/dynamic-rhythms/starter-notebook/utils.py

```diff
@@ -0,0 +1,708 @@
+    file_name = f"eaglei_outages_{year}.csv"
+    file_path = os.path.join(data_directory, file_name)
+    
+    # Check if the file exists
+    if not os.path.exists(file_path):
+        raise FileNotFoundError(f"The file {file_name} does not exist in the directory {data_directory}.")
+    
+    # Read the CSV file into a DataFrame
+    df = pd.read_csv(file_path)
+    df['run_start_time'] = pd.to_datetime(df['run_start_time'])
+    df.dropna(subset=['customers_out'],inplace=True)
+    
+    
+    url = "https://www2.census.gov/geo/tiger/GENZ2021/shp/cb_2021_us_county_20m.zip"
+    counties = gpd.read_file(url)
+
+    counties = counties[counties['STATE_NAME'] == state]
+    
+
+    # Filter data based on time range and state
+    filtered_df = df[(df['run_start_time'] >= start_time) & (df['run_start_time'] <= end_time) & (df['state'] == state)]
+    
+    grouped = filtered_df.groupby(['county', 'fips_code']).agg({'customers_out': 'sum'}).reset_index()
+    grouped['customers_out'] = grouped['customers_out'] / 96
+    
+    
+    # Merge the Texas counties with aggregated data
+    grouped['fips_code'] = grouped['fips_code'].astype(str)  # Ensure FIPS codes are strings
+    counties = counties.merge(grouped, left_on='NAME', right_on='county', how='left')
+    
+    # Fill missing values with zero so we display it in white color
+    counties['customers_out'] = counties['customers_out'].fillna(0) 
+
+
+    # Create a custom colormap
+    viridis_cmap = plt.cm.viridis  # Base colormap
+    colors = viridis_cmap(np.linspace(0, 1, 256))  # Extract colors from viridis
+    colors[0] = np.array([1, 1, 1, 1])  # Set the first color (for zero values) to white
+    custom_cmap = ListedColormap(colors)
+    
+    # Plot the map
+    fig, ax = plt.subplots(1, 1, figsize=(10, 6))
+    county_plot = counties.plot(column='customers_out',  # Data to visualize
+                                cmap=custom_cmap,  # Colormap
+                                linewidth=0.8,
+                                ax=ax,
+                                edgecolor='black',
+                                legend=False,  # Disable automatic colorbar
+                                norm=Normalize(vmin=0, vmax=counties['customers_out'].max())  # Normalize scale
+                               )
+    
+    # Add colorbar and set its title
+    sm = plt.cm.ScalarMappable(cmap=custom_cmap, norm=Normalize(vmin=0, vmax=counties['customers_out'].max()))
+    sm._A = []  # Dummy array for the ScalarMappable
+    cbar = fig.colorbar(sm, ax=ax)
+    # cbar.set_label('Number of Customers Out', fontsize=12, labelpad=10)
+    cbar.ax.xaxis.set_label_position('top')
+    cbar.ax.xaxis.set_ticks_position('top')
+    cbar.ax.set_xlabel('customer.day', fontsize=14)
+    
+    # Remove the vertical label
+    cbar.ax.set_ylabel('')
+    
+    # Customize the plot
+    title = f'Power outage in {state}\nfrom {start_time.strftime('%Y-%m-%d')} to {end_time.strftime('%Y-%m-%d')}'
```

Comment:
not sure why this is throwing `error: cannot format utils.py: cannot use --safe with this file; failed to parse source file AST: f-string: unmatched '(' (<unknown>, line 222)` when I try to reformat with black